qcodes/instrument_drivers/stanford_research/SIM928.py

```diff
@@ -0,0 +1,264 @@
+                           parameter_class=ManualParameter,
+                           vals=vals.Numbers(0,1), initial_value=0.05)
+
+        super().connect_message()
+
+    def get_idn(self, i=None):
+        """
+        i:
+            Slot of the module whose id is returned
+        Returns:
+            A dict containing vendor, model, serial, and firmware.
+        """
+        try:
+            idstr = ''  # in case self.ask fails
+            if i is None:
+                idstr = self.ask('*IDN?')
+            else:
+                idstr = self.ask_module(i, '*IDN?')
+            # form is supposed to be comma-separated, but we've seen
+            # other separators occasionally
+            for separator in ',;:':
+                # split into no more than 4 parts, so we don't lose info
+                idparts = [p.strip() for p in idstr.split(separator, 3)]
+                if len(idparts) > 1:
+                    break
+            # in case parts at the end are missing, fill in None
+            if len(idparts) < 4:
+                idparts += [None] * (4 - len(idparts))
+        except:
+            logging.warning('Error getting or interpreting *IDN?: ' +
+                            repr(idstr))
+            idparts = [None, None, None, None]
+
+        # some strings include the word 'model' at the front of model
+        if str(idparts[1]).lower().startswith('model'):
+            idparts[1] = str(idparts[1])[5:].strip()
+
+        return dict(zip(('vendor', 'model', 'serial', 'firmware'), idparts))
+
+    def find_modules(self):
+        """
+        Queries the SIM900 mainframe, in which slots there is a module present.
+        :return: a list of slot numbers where a module is present (starting
+                 from 1)
+        """
+        CTCR = self.ask('CTCR?')
+        CTCR = int(CTCR) >> 1
+        modules = []
+        for i in range(1,10):
+            if CTCR & 1 != 0:
+                modules.append(i)
+            CTCR >>= 1
+        return modules
+
+    def ask_module(self, i, cmd):
+        msg = 'SNDT {},"{}"'.format(i, cmd)
+        self.write(msg)
+        time.sleep(100e-3)
+        msg = 'GETN? {},128'.format(i)
+        msg = self.ask(msg)
+        # first read consumes the terminator of the message from the submodule,
+        # so we have a terminator from the message to us still in the input
+        # buffer.
+        self.visa_handle.read()
+
+        if msg[:2] != '#3':
+            raise RuntimeError('Unexpected format of answer: {}'.format(msg))
+        return msg[5:]
+
+    def write_module(self, i, cmd):
+        self.write('SNDT {},"{}"'.format(i, cmd))
+
+    def set_voltage(self, i, voltage):
+        self.write_module(i, 'VOLT {:.3f}'.format(voltage))
+        self.parameters['voltage{}'.format(i)]._save_val(voltage)
+
+    def get_voltage(self, i):
+        return float(self.ask_module(i, 'VOLT?'))
+
+    def set_smooth(self, voltagedict, equitime=False, verbose=False):
```

Comment:
This is a very useful function! We usually give the SIM modules a name, such as `top_gate`, and then forget about the slot number it is in. Would it make sense for the voltagedict to also allow parameter names as keys? It can then retrieve the corresponding slot?